### ISSUE TITLE
Refactor: Improve no-magic-numbers const detection robustness

### ISSUE THREAD

**Issue by pertrai1** [pertrai1/eslint-plugin-llm-core#39]
## Description

The `isConstAssignment` function in `no-magic-numbers.ts` uses a manual AST traversal that could be simplified and made more robust.

## Current Issues

1. Manual traversal with `node = current; current = current.parent` is fragile
2. Doesn't handle nested expressions well (e.g., `const X = fn(5)` or `const X = arr[5]`)
3. The logic is hard to follow and maintain

## Suggested Refactor

Use a helper function that traces ancestry to find if the number is the direct init value of a const declarator:

```ts
function isDirectConstInit(node: TSESTree.Literal): boolean {
  // Walk up until we find a VariableDeclarator
  // Check that: (1) it's a const, (2) the number is the direct init (not nested in call/array access)
}
```

Add more tests for edge cases:
- `const X = 5 + 3` (binary expr - should allow)
- `const X = fn(5)` (call argument - should NOT allow)
- `const X = arr[5]` (array index - should NOT allow) 
- `const X = { val: 5 }` (object value - should NOT allow)
- `const { x = 5 } = obj` (default value - handled separately)
